internal/metrics/sources/prometheus/prometheus.go

```diff
@@ -0,0 +1,265 @@
+	return model.Prometheus
+}
+
+// SplitN arguments are dependent on what Prometheus data looks like for each type of line.
+func scrapeEndpoint(ctx context.Context, target string) ([]model.DataEntry, error) {
+	rows, err := fetchPrometheusData(ctx, target)
+	if err != nil {
+		return nil, fmt.Errorf("failed to query Prometheus metrics endpoint: %w", err)
+	}
+
+	entries := make([]model.DataEntry, 0)
+	builder := model.NewEntryBuilder(model.WithSourceType(model.Prometheus))
+	for _, row := range rows {
+		entries = processRow(row, &builder, entries)
+	}
+
+	// Add the final entry being parsed, as the loop above adds entries only on '# HELP' lines, and the final line is
+	// NOT a '# HELP' line
+	if builder.CanBuild() {
+		entries = append(entries, builder.Build())
+	}
+
+	slog.Debug("Prometheus endpoint scraped", "number of data entries", len(entries), "Prometheus URL", target)
+
+	return entries, nil
+}
+
+func fetchPrometheusData(ctx context.Context, targetURL string) ([]string, error) {
+	responseString, err := queryTarget(ctx, targetURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return strings.Split(responseString, "\n"), nil
+}
+
+// queries the given URL and returns its content as a string.
+func queryTarget(ctx context.Context, url string) (string, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return "", fmt.Errorf("failed to formulate Prometheus HTTP scrape endpoint GET request: %w", err)
+	}
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("GET request to Prometheus HTTP scrape endpoint [%s] failed: %w", url, err)
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse Prometheus scrape endpoint response body: %w", err)
+	}
+
+	err = resp.Body.Close()
+	if err != nil {
+		return "", fmt.Errorf("failed to close response body: %w", err)
+	}
+
+	return string(body), nil
+}
+
+const (
+	// Prefixes that label the content of Prometheus data rows.
+	prometheusHelpPrefix = "# HELP"
+	prometheusTypePrefix = "# TYPE"
+)
+
+func processRow(row string, builder *model.EntryBuilder, entries []model.DataEntry) []model.DataEntry {
+	result := entries
+	if strings.HasPrefix(row, prometheusHelpPrefix) {
+		result = appendEntry(builder, entries)
+		processHelpRow(row, builder)
+	} else if strings.HasPrefix(row, prometheusTypePrefix) {
+		processTypeRow(row, builder)
+	} else if row != "" {
+		processPointRow(row, builder)
+	}
+
+	return result
+}
+
+// Appends the contents of the builder into the result slice if builder is non-empty.
+func appendEntry(builder *model.EntryBuilder, entries []model.DataEntry) []model.DataEntry {
+	result := entries
+	// If entry already has content, then we add it to the result slice and reset.
+	if builder.CanBuild() {
+		result = append(result, builder.Build())
+		*builder = model.NewEntryBuilder(model.WithSourceType(model.Prometheus))
+	}
+
+	return result
+}
+
+const (
```

Comment:
same comments about the consts here, probably more readable at the top of the file

Reply:
https://github.com/nginx/agent/pull/544#discussion_r1483071548